poll/poll.py

```diff
@@ -300,6 +316,89 @@ def wrapper(self, request_json, suffix=''):
         # Otherwise fall back to a standard JSON handler
         return cls.json_handler(func)
 
+    def _get_status(self):
+        self.check_pending_export()
+        return {
+            'export_pending': bool(self.active_export_task_id),
+            'last_export_result': self.last_export_result,
+            'download_url': self.download_url_for_last_report,
+        }
+
+    def check_pending_export(self):
+        """
+        If we're waiting for an export, see if it has finished, and if so, get the result.
+        """
+        from .tasks import export_data as export_data_task  # Import here since this is edX LMS specific
+        if self.active_export_task_id:
+            async_result = export_data_task.AsyncResult(self.active_export_task_id)
+            if async_result.ready():
+                self._store_export_result(async_result)
+
+    @property
+    def download_url_for_last_report(self):
+        """ Get the URL for the last report, if any """
+        # Unfortunately this is a bit inefficient due to the ReportStore API
+        if not self.last_export_result or self.last_export_result['error'] is not None:
+            return None
+
+        report_store = ReportStore.from_config(config_name='GRADES_DOWNLOAD')
+        course_key = getattr(self.scope_ids.usage_id, 'course_key', None)
+        return dict(report_store.links_for(course_key)).get(self.last_export_result['report_filename'])
+
+    def student_module_queryset(self):
+        return StudentModule.objects.filter(
+            course_id=self.runtime.course_id,
+            module_state_key=self.scope_ids.usage_id,
+        ).order_by('-modified')
+
+    @XBlock.json_handler
+    def csv_export(self, data, suffix=''):
+        """
+        Asynchronously export given data as a CSV file.
+        """
+        # Launch task
+        from .tasks import export_data as export_data_task  # Import here since this is edX LMS specific
+
+        # Make sure we nail down our state before sending off an asynchronous task.
+        async_result = export_data_task.delay(
+            self.prepare_data(),
```

Comment:
By calling `prepare_data()` here, most of the work is being done synchronously, before we even queue the asynchronous task. All that the asynchronous task is doing is writing the results to a CSV file, which is a trivial task that's not worth making asynchronous.

You should instead be calling `export_data_task` with only one argument (the block ID), and within that task is where we should be preparing the data, which is the part that we want to make asynchronous.